### ISSUE TITLE
Increase a default value of reader_concurrency_semaphore_cpu_concurrency (currently 1)

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/scylladb#21335]
HEAD: reader_concurrency_semaphore_cpu_concurrency

## Description
Current default value of `reader_concurrency_semaphore_cpu_concurrency` effectively allows no concurrency for requests that don't require I/O which makes little to no sense:
* For requests that don't preempt this configurable doesn't have any impact.
* For cpu-needing requests that do preempt (e.g. due to memory constraints or due to running out of task-quota runtime) allowing other shorter cpu-needing requests makes a lot of sense allowing processing requests in parallel.

Increasing the default `reader_concurrency_semaphore_cpu_concurrency` so something like 100 (mimicking a maximum number of I/O-needing requests limit) would only make things better performance-wise.

**Comment by vladzcloudius** [comment]
> This effectively means removing the CPU concurrency altogether. I added this feature on request from [@avikivity](https://github.com/avikivity) so we (I mean you) need to convince him first that we don't really need CPU concurrency control.

It's the other way around actually, @denesb - having it set to 1 effectively means "no CPU concurrency". And I'm advocating to **do** allow a reasonable amount of CPU concurrency.

I put a few bullets in the opening message explaining why allowing concurrent "CPU-needing" requests can only do good.

Allowing internal concurrency in general is a pretty obvious good practice so I'm not sure what needs convincing here. We had enough real-life cases so far were `reader_concurrency_semaphore_cpu_concurrency: 1` was a bad choice. I'm not familiar with any case were `100 > reader_concurrency_semaphore_cpu_concurrency > 1` was a bad choice. Are you familiar with any, @denesb @avikivity?

@avikivity if you could clarify what good `reader_concurrency_semaphore_cpu_concurrency: 1` makes us that would make it easier for me to "convince" you.

**Comment by dorlaor** [comment]
Can you demonstrate the simplest case which this help us? it will be harder to argue with facts

**Comment by vladzcloudius** [comment]
> Can you demonstrate the simplest case which this help us? it will be harder to argue with facts

Sure.
Long query that reads from cache only (e.g. large range scan) running in parallel with small reads from cache.
The long read takes a few reactor task quotas while a single task quota is enough for multiple small reads to complete.

With the current defaults small reads are going to be starved by the large query and get its latency as an added latency to their latency.

With a larger CPU concurrency small reads can enter and get completed a lot sooner.

The latency of the large read would get impacted but latency for such reads is large by definition so that added latency is not going to be noticed.

I'm pretty sure this example should be very familiar to you, @dorlaor ;)

**Comment by denesb** [comment]
> > This effectively means removing the CPU concurrency altogether. I added this feature on request from [@avikivity](https://github.com/avikivity) so we (I mean you) need to convince him first that we don't really need CPU concurrency control.
> 
> It's the other way around actually, [@denesb](https://github.com/denesb) - having it set to 1 effectively means "no CPU concurrency". And I'm advocating to **do** allow a reasonable amount of CPU concurrency.

Sorry, I meant removing the CPU concurrency control feature from the semaphore altogether.

**Comment by denesb** [comment]
@vladzcloudius how many clusters are using `reader_concurrency_semaphore_cpu_concurrency=100`? Have you observed any adverse effects when increasing `reader_concurrency_semaphore_cpu_concurrency` in any cluster?

**Comment by denesb** [comment]
> Allowing internal concurrency in general is a pretty obvious good practice so I'm not sure what needs convincing here. We had enough real-life cases so far were `reader_concurrency_semaphore_cpu_concurrency: 1` was a bad choice. I'm not familiar with any case were `100 > reader_concurrency_semaphore_cpu_concurrency > 1` was a bad choice. Are you familiar with any, [@denesb](https://github.com/denesb) [@avikivity](https://github.com/avikivity)?

I do not have any cases, either pro or contra. Actually, one contra, the Discord case, which initiated the introduction of this config parameter in the first place.
This was the first complaint since this feature was implemented, but this is not really proof in any way. I think for most workloads, the feature is neutral and has no effect -- if requests are small enough, they will finish without yield anyway.

**Comment by vladzcloudius** [comment]
> Sorry, I meant removing the CPU concurrency control feature from the semaphore altogether.

I wasn't suggesting doing it either. Removing the concurrency control completely is unwise IMO - we added it for a good reason originally. I'm only suggesting to increase the default value.

> [@vladzcloudius](https://github.com/vladzcloudius) how many clusters are using `reader_concurrency_semaphore_cpu_concurrency=100`? Have you observed any adverse effects when increasing `reader_concurrency_semaphore_cpu_concurrency` in any cluster?

I know about 3 cases (see issues that reference this one) and only "good effects" came out of it. ;)
@fee-mendes @mkeeneyj  do you have any additional stats?

**Comment by fee-mendes** [comment]
@denesb Enterprise-specific customer names should not be under this venue.

**Comment by fee-mendes** [comment]
Unless anything changed since when I first stumbled upon this, one reason to simply bump the limit from 1 was the fact that a single long-running query can prevent clients from connecting to the cluster. 

I've actually demonstrated that in one of the previous issues, and simply bumping the concurrency allowed clients in.

**Comment by denesb** [comment]
> [@denesb](https://github.com/denesb) Enterprise-specific customer names should not be under this venue.

Sorry, I though this is an enterprise issue (not sure why). I revised my comment and dropped the revision with the customer mention.

**Comment by mykaul** [comment]
@denesb - please assign. Somehow it missed 2025.1.0?

**Comment by denesb** [comment]
This was already done by https://github.com/scylladb/scylladb/commit/3d12451d1f6c160c64228c378a8811e6945bc28c, and it is part of 2025.1.0.
I guess we can close this, if @vladzcloudius is happy with the new default of 2 although note that the new default was not backported beyond 2025.1.

**Comment by mykaul** [comment]
> This was already done by [3d12451](https://github.com/scylladb/scylladb/commit/3d12451d1f6c160c64228c378a8811e6945bc28c), and it is part of 2025.1.0. I guess we can close this, if [@vladzcloudius](https://github.com/vladzcloudius) is happy with the new default of 2 although note that the new default was not backported beyond 2025.1.

It shouldn't be backported to earlier releases.
